src/linker/Linker.Dataflow/DynamicallyAccessedMembersTypeHierarchy.cs

```diff
@@ -216,13 +216,63 @@ public DynamicallyAccessedMembersTypeHierarchy (LinkContext context, MarkStep ma
+		{
+			Debug.Assert (annotation != DynamicallyAccessedMemberTypes.None);
+
+			if (type.Name == "AnnotatedAll")
+				System.Console.WriteLine ("Hey");
+
+			reflectionPatternContext.AnalyzingPattern ();
+			// Handle cases where a type has no members but annotations are to be applied to derived type members
+			reflectionPatternContext.RecordHandledPattern ();
+
+			// We need to apply annotations to this type, and its base/interface types (recursively)
+			// But the annotations on base/interfaces are already applied so we don't need to apply those
+			// again (and should avoid doing so as it would produce extra warnings).
+			var baseType = _context.TryResolve (type.BaseType);
+			if (baseType != null) {
+				var baseAnnotation = GetCachedInfoForTypeInHierarchy (baseType);
+				var annotationToApplyToBase = ReflectionMethodBodyScanner.GetMissingMemberTypes (annotation, baseAnnotation.annotation);
+
+				// Apply any annotations that didn't exist on the base type to the base type.
+				// This may produce redundant warnings when the annotation is DAMT.All or DAMT.PublicConstructors and the base already has a
+				// subset of those annotations.
+				reflectionMethodBodyScanner.MarkTypeForDynamicallyAccessedMembers (ref reflectionPatternContext, baseType, annotationToApplyToBase, DependencyKind.DynamicallyAccessedMemberOnType, declaredOnly: false);
+			}
+
+			// Most of the DynamicallyAccessedMemberTypes don't select members on interfaces. We only need to apply
+			// annotations to interfaces separately if dealing with DAMT.All or DAMT.Interfaces.
+			if (annotation.HasFlag (DynamicallyAccessedMemberTypesOverlay.Interfaces) && type.HasInterfaces) {
+				var annotationToApplyToInterfaces = annotation == DynamicallyAccessedMemberTypes.All ? annotation : DynamicallyAccessedMemberTypesOverlay.Interfaces;
+				foreach (var iface in type.Interfaces) {
+					var interfaceType = _context.TryResolve (iface.InterfaceType);
+					if (interfaceType == null)
+						continue;
+
+					var interfaceAnnotation = GetCachedInfoForTypeInHierarchy (interfaceType);
+					if (interfaceAnnotation.annotation.HasFlag (annotationToApplyToInterfaces))
+						continue;
+
+					// Apply All or Interfaces to the interface type.
+					// DAMT.All may produce redundant warnings from implementing types, when the interface type already had some annotations.
+					reflectionMethodBodyScanner.MarkTypeForDynamicallyAccessedMembers (ref reflectionPatternContext, interfaceType, annotationToApplyToInterfaces, DependencyKind.DynamicallyAccessedMemberOnType, declaredOnly: false);
+				}
+			}
+
+			// The annotations this type inherited from its base types or interfaces should not produce
+			// warnings on the respective base/interface members, since those are already covered by applying
+			// the annotations to those types.
```

Comment:
```suggestion
			// The annotations this type inherited from its base types or interfaces should not produce
			// warnings on the respective base/interface members, since those are already covered by applying
			// the annotations to those types. So we only need to handle the members directly declared on this type.
```